src/lib/updating-element.ts

```diff
@@ -468,69 +480,83 @@ export abstract class UpdatingElement extends HTMLElement {
    * @returns {Promise} A Promise that is resolved when the update completes.
    */
   requestUpdate(name?: PropertyKey, oldValue?: any) {
+    let shouldRequestUpdate = true;
+    // if we have a property key, perform property update steps.
     if (name !== undefined) {
-      const options = (this.constructor as typeof UpdatingElement)
-                          ._classProperties.get(name) ||
-                      defaultPropertyDeclaration;
-      return this._requestPropertyUpdate(name, oldValue, options);
+      const ctor = this.constructor as typeof UpdatingElement;
+      const options =
+          ctor._classProperties.get(name) || defaultPropertyDeclaration;
+      if (ctor._valueHasChanged(this[name as keyof this], oldValue,
+                                options.hasChanged)) {
+        // track old value when changing.
+        if (!this._changedProperties.has(name)) {
+          this._changedProperties.set(name, oldValue);
+        }
+        // add to reflecting properties set
+        if (options.reflect === true) {
+          if (this._reflectingProperties === undefined) {
+            this._reflectingProperties = new Map();
+          }
+          this._reflectingProperties.set(name, options);
+        }
+        // abort the request if the property should not be considered changed.
+      } else {
+        shouldRequestUpdate = false;
+      }
     }
-    return this._invalidate();
+    if (!this._hasRequestedUpdate && shouldRequestUpdate) {
+      this._setupUpdate();
+    }
+    return this.updateComplete;
   }
 
   /**
-   * Requests an update for a specific property and records change information.
-   * @param name {PropertyKey} name of requesting property
-   * @param oldValue {any} old value of requesting property
-   * @param options {PropertyDeclaration}
-   */
-  private _requestPropertyUpdate(name: PropertyKey, oldValue: any,
-                                 options: PropertyDeclaration) {
-    if (!(this.constructor as typeof UpdatingElement)
-             ._valueHasChanged(this[name as keyof this], oldValue,
-                               options.hasChanged)) {
-      return this.updateComplete;
+   * Sets up the element to asynchronously update.
+   */
+  private async _setupUpdate() {
+    // Mark state updating...
+    this._updateState = this._updateState | STATE_UPDATE_REQUESTED;
+    let resolve: (r: boolean) => void;
+    const previousUpdatePromise = this._updatePromise;
+    this._updatePromise = new Promise((res) => resolve = res);
+    // Ensure any previous update has resolved before updating.
+    // This `await` also ensures that property changes are batched.
+    await previousUpdatePromise;
+    // Make sure the element has connected before updating.
+    if (!this._hasConnected) {
+      await new Promise((res) => this._hasConnectedResolver = res);
```

Comment:
It seems unfortunate to need to introduce another promise to synchronize a pending update with connected, but I get that it is needed to keep the `updateComplete` promise pending to allow someone to `await el.updateComplete` before it is connected.  I don't have a better idea, other than saving the `_updatePromise` resolver and letting connected's call to `requestUpdate` use the same promise.  Not sure that's any better.

Would it be too weird to just document that the pending update system isn't ready until first connected, so `updateComplete` shouldn't be awaited until the element is connected?  That would allow you to just bail in `requestUpdate` (since you'll always come back in when connected).

If you don't like either of those suggestions this code is fine, not that big of a deal.  Just somehow feels more complicated than it should be and wanted to brainstorm a little.

Reply:
Acknowledged, but think it makes sense to keep as is.